test/e2e/instrumentation/monitoring/prometheus.go

```diff
@@ -0,0 +1,389 @@
+		consumer := exportCustomMetric(f, consumerName, int(customMetricTargetValue), annotatePods, annotateService)
+		defer consumer.CleanUp()
+		retryUntil(func() error {
+			return validateQueryReturnsCorrectValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, prometheusTestTimeout)
+	})
+})
+
+func prometheusCPUQuery(namespace, podNamePrefix string, rate time.Duration) string {
+	return fmt.Sprintf(`sum(irate(container_cpu_usage_seconds_total{namespace="%v",pod_name=~"%v.*"}[%vm]))`,
+		namespace, podNamePrefix, int64(rate.Minutes()))
+}
+
+func prometheusServiceCustomMetricQuery(namespace, service string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%v",kubernetes_name="%v"})`, namespace, service)
+}
+
+func prometheusPodCustomMetricQuery(namespace, podNamePrefix string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%s",kubernetes_pod_name=~"%s.*"})`, namespace, podNamePrefix)
+}
+
+func consumeCPUResources(f *framework.Framework, consumerName string, cpuUsage int) *common.ResourceConsumer {
+	return common.NewDynamicResourceConsumer(consumerName, f.Namespace.Name, common.KindDeployment, 1, cpuUsage,
+		memoryUsed, 0, int64(cpuUsage), memoryLimit, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func exportCustomMetric(f *framework.Framework, consumerName string, metricValue int, annotatePods, annotateService bool) *common.ResourceConsumer {
+	var podAnnotations, serviceAnnotations map[string]string
+	if annotatePods {
+		podAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	if annotateService {
+		serviceAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	return common.NewMetricExporter(consumerName, f.Namespace.Name, podAnnotations, serviceAnnotations, metricValue, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func validateMetricAvailableForNodes(c clientset.Interface, metric string, expectedNodesNames []string) error {
+	instanceLabels, err := getInstanceLabelsAvailableForMetric(c, prometheusRequiredNodesUpDuration, metric)
+	if err != nil {
+		return err
+	}
+	nodesWithMetric := make(map[string]bool)
+	for _, instance := range instanceLabels {
+		nodesWithMetric[instance] = true
+	}
+
+	for _, nodeName := range expectedNodesNames {
+		if _, found := nodesWithMetric[nodeName]; !found {
+			return fmt.Errorf("Metric '%v' is not available for node '%v'", metric, nodeName)
+		}
+	}
+	return nil
+}
+
+func validateAllActiveTargetsAreHealthy(c clientset.Interface) error {
+	discovery, err := fetchPrometheusTargetDiscovery(c)
+	if err != nil {
+		return err
+	}
+	if len(discovery.ActiveTargets) == 0 {
+		return fmt.Errorf("Prometheus is not scraping any targets, at least one target is required")
```

Comment:
What targets are required? I don't ask you do change code here, just want to understand it better. E.g. is a system pod an active target? Why do we expect all targets to be healthy, will it be reported unhealthy if some system component is failing? Should it cause Prometheus tests to fail?

Reply:
Target is url endpoint that is scraped by prometheus. It can be anything that is found by prometheus service discovery or hardcoded (nodes, pods, services, ingress, apiserver). Prometheus applies rules on those targets that can modify them, add labels or discard them. Active targets are those that were not discarded. For kubernetes we will drop all targets that do not have proper annotations. 

Not healthy targets can mean that prometheus didn't manage to run request (there is delay until requests are made or prometheus wasn't fast enough to make request) or request failed (misconfiguration, timeout, unauthorized).

In that test I want mainly verify configuration, that all targets are reachable.

Reply:
Got it, this is very informative, consider adding this explanation (or part of it) in the code comments as well.

I'm also thinking if it's possible to hit a window when only one or few targets are discovered and are succeeding, and we didn't get to test some failing targets. Do you think it makes sense to wait for some more representative set of active targets to test? (Not necessarily in this PR)

Reply:
Some targets are available from configuration (static prometheus self monitoring) and other are depended on service discovery that requires call to kubernetes apiserver. I would guess that there is window between reloading configuration and apiserver response (especially for large clusters).

We don't want to pin to specific job, but we may want to ensure that targets were discovered and will not change.

Reply:
Sounds good, something like checking that there is the same number of targets between two iterations before proceeding?

Reply:
I was thinking about something right that with small percent tolerations to handle large clusters.